internal/overlord/servstate/manager.go

```diff
@@ -477,3 +477,74 @@ func (m *ServiceManager) CheckFailed(name string) {
 		}
 	}
 }
+
+var shutdownTimeout = killWait + 100*time.Millisecond
+
+// Shutdown shuts down the service manager, stopping all running services.
+func (m *ServiceManager) Shutdown() error {
+	services, err := m.servicesToStop()
+	if err != nil {
+		return err
+	}
+	if len(services) == 0 {
+		logger.Debugf("No services to stop.")
+		return nil
+	}
+
+	// One change to stop them all.
+	logger.Noticef("Stopping all running services.")
+	st := m.state
+	st.Lock()
+	taskSet, err := Stop(st, services)
+	if err != nil {
+		st.Unlock()
+		return err
+	}
+	chg := st.NewChange("shutdown", "Shut down service manager")
+	chg.AddAll(taskSet)
+	chg.Set("service-names", services)
+	st.EnsureBefore(0)
+	st.Unlock()
+
+	// Wait for a limited amount of time for them to stop.
+	select {
+	case <-chg.Ready():
+		logger.Debugf("All services stopped.")
+	case <-time.After(shutdownTimeout):
+		return fmt.Errorf("timed out after %v", shutdownTimeout)
+	}
```

Comment:
Thanks. Yeah, the convention is the caller would add that context, so for example:

```go
err := Shutdown()
if err != nil {
    return fmt.Errorf("cannot shut down service manager: %w", err)
}
```

In actual fact for this case we log the shutdown error and add context in the log mesage. Here's the actual calling code:

```go
	err := d.overlord.ServiceManager().Shutdown()
	if err != nil {
		// This isn't fatal for exiting the daemon, so log and continue.
		logger.Noticef("Cannot shut down service manager: %v", err)
	}
```

So it would print something like:

```
2022-04-28T22:27:32.261Z [pebble] Cannot shut down service manager: timed out after 5.1s
```

Reply:
There's still a point about this being a complex operation, though, and the call site has no clue of what actually timed out. So maybe:

`"timeout waiting for change to complete"`

... would give some helpful hint.

Reply:
Changed (in the refactoring) to "timeout stopping running services".